### DISCUSSION TITLE
Profiling a Workflow

### DISCUSSION THREAD

**Discussion by ni-todo-spot** [argoproj/argo-workflows#12029] [Q&A] [answered]
Hello all!

I'm interested in profiling a specific workflow template (ie its corresponding submissions), so I can find slow code and bottlenecks.

I was wondering if there is some builtin feature, some kind of documentation and generally what are my options.

Thanks in advance 🙏

**Comment by agilgur5** [comment, marked as answer]
To monitor a running workflow, you can use [`metrics`](https://argoproj.github.io/argo-workflows/variables/#metrics) to output information about a particular step/task, which a metrics provider (e.g. Prometheus) can scrape.

_Profiling_ the code itself is a bit different though, that requires your code to be instrumented. If your code is instrumented, it will typically send profiling and tracing data to a collector.
That is outside the scope of Argo, and Argo can't instrument code for you (Argo just has your container image). Instrumentation is the same for any piece of code and not specific to Argo either. It often requires a vendor-specific library and configuring a collector specifically as well. As an example, here are docs for [adding NewRelic APM to Python tasks](https://docs.newrelic.com/docs/apm/agents/python-agent/getting-started/introduction-new-relic-python/#non-web).
